apis/python/src/tiledbsoma/io/shaping.py

```diff
@@ -0,0 +1,546 @@
+# Copyright (c) 2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+"""
+Provides support for the new shape feature in TileDB-SOMA 1.15, including the
+ability to process all dataframes/arrays contained within a TileDB-SOMA
+Experiment. Please also see
+https://github.com/single-cell-data/TileDB-SOMA/issues/2407.  """
+
+import io
+import sys
+from typing import Any, Dict, Optional, Tuple, TypedDict, Union, cast
+
+import tiledbsoma
+
+Printable = Union[io.TextIOWrapper, io.StringIO]
+
+
+class SizingArgs(TypedDict):
+    """Convenience type-alias for kwargs passed to experiment-level
+    upgrade/resize functions.
+    """
+
+    nobs: Optional[int]
+    nvars: Optional[Dict[str, int]]
+    ms_name: Optional[str]
+    coll_name: Optional[str]
+    verbose: bool
+    check_only: bool
+    context: Optional[tiledbsoma.SOMATileDBContext]
+    output_handle: Printable
```

Comment:
Would suggest importing the `attrs` module for using data classes. You can leverage default values with this.

```suggestion
@attr.s(auto_attribs=True, frozen=True)
class SizingArgs:
    nobs: Optional[int] = None
    nvars: Optional[Dict[str, int]] = None
    ms_name: Optional[str] = None
    coll_name: Optional[str] = None
    verbose: bool = True
    check_only: bool = True
    context: Optional[tiledbsoma.SOMATileDBContext] = None
    output_handle: Printable
```